### PR TITLE
[Feature] Replace AES256GCM with ChaCha20

### PR DESCRIPTION
**Pull request by johnxnguyen**
## What's new in this PR?

This PR replaces the AES256GCM encryption wrapper with the XChaCha20Poly1305 encryption wrapper. Unfortunately, AES256GCM is not available on ARM processors. Fortunately, the constructions for both encryption algorithms are the same and it is simply a matter or changing the underlying sodium method calls.

To avoid naming conflicts with the existing `ChaCha20Encryption` wrapper, which is a stream based wrapper, I introduce the `ChaCha20` namespace and placed the stream wrapper under `StreamEncryption` and the AEAD wrapper under `AEADEncryption`.

### CODE REVIEW

File: WireCryptobox/ChaCha20/AEADEncryption.swift [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+            let contextBytes = context.bytes
+            let contextLength = UInt64(contextBytes.count)
+
+            let nonceBytes = generateRandomNonceBytes()
+
+            var ciphertextBytes = createByteArray(length: ciphertextLength(forMessageLength: Int(messageLength)))
+            var actualCiphertextLength: UInt64 = 0
+
+            crypto_aead_xchacha20poly1305_ietf_encrypt(
```
Comment (johnxnguyen) [inline WireCryptobox/ChaCha20/AEADEncryption.swift:57]: This changed from `crypto_aead_aes256gcm_encrypt`.

File: WireCryptobox/ChaCha20/AEADEncryption.swift [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            let ciphertextBytes = ciphertext.bytes
+            let ciphertextLength = UInt64(ciphertextBytes.count)
+
+            let contextBytes = context.bytes
+            let contextLength = UInt64(contextBytes.count)
+
+            var messageBytes = createByteArray(length: messageLength(forCiphertextLength: Int(ciphertextLength)))
+            var actualMessageLength: UInt64 = 0
+
+            let result = crypto_aead_xchacha20poly1305_ietf_decrypt(
```
Comment (johnxnguyen) [inline WireCryptobox/ChaCha20/AEADEncryption.swift:102]: This changed from `crypto_aead_aes256gcm_decrypt`.

File: WireCryptobox/ChaCha20/AEADEncryption.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            guard length == messageLength + UInt64(authenticationBytesLength) else {
+                throw EncryptionError.malformedCiphertext
+            }
+        }
+
+        // MARK: - Buffer Lengths
+
+        private static let keyLength = Int(crypto_aead_xchacha20poly1305_ietf_KEYBYTES)
+        private static let nonceLength = Int(crypto_aead_xchacha20poly1305_ietf_NPUBBYTES)
+        private static let authenticationBytesLength = Int(crypto_aead_xchacha20poly1305_ietf_ABYTES)
```
Comment (johnxnguyen) [inline WireCryptobox/ChaCha20/AEADEncryption.swift:170]: These changed from `crypto_aead_aes256gcm_*`.

**Comment by johnxnguyen** [comment]
@franziskuskiefer unfortunately the diff doesn't show the files as being renamed but as deleted and new ones added. I didn't change the implementation of the `ChaCha20Encryption` class at all (just renamed to `StreamEncryption` and moved it. And for the `AEADEncryption`, I left comments on the lines that changed, otherwise the rest of the implementation is unchanged.

File: WireCryptobox/ChaCha20/AEADEncryption.swift [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public extension ChaCha20 {
```
Comment (franziskuskiefer) [inline WireCryptobox/ChaCha20/AEADEncryption.swift:21]: Please call it `ChaCha20Poly1305`. It's actually a different thing.

File: WireCryptobox/ChaCha20/AEADEncryption.swift [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+    /// AEAD Encryption wrapper for ChaCha20.
```
Comment (franziskuskiefer) [inline WireCryptobox/ChaCha20/AEADEncryption.swift]: "for ChaCha20Poly1305" again.
ChaCha20 is the encryption (like AES) and Poly1305 is the authentication (like GCM).

**Comment by franziskuskiefer** [comment]
> @franziskuskiefer unfortunately the diff doesn't show the files as being renamed but as deleted and new ones added. I didn't change the implementation of the `ChaCha20Encryption` class at all (just renamed to `StreamEncryption` and moved it. And for the `AEADEncryption`, I left comments on the lines that changed, otherwise the rest of the implementation is unchanged.

Ah, I didn't see the comment :) But that's what I assumed.

File: WireCryptobox/ChaCha20/AEADEncryption.swift (hunk 5 above)
Comment (johnxnguyen) [reply WireCryptobox/ChaCha20/AEADEncryption.swift]: Ok, thanks for the explanation!

**Comment by johnxnguyen** [comment]
@zenkins release major

**Pull request merged**